core/src/kvs/surrealkv/mod.rs

```diff
@@ -0,0 +1,398 @@
+		self.check = check;
+	}
+	/// Check if closed
+	pub(crate) fn closed(&self) -> bool {
+		self.done
+	}
+	/// Cancel a transaction
+	pub(crate) async fn cancel(&mut self) -> Result<(), Error> {
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Mark this transaction as done
+		self.done = true;
+
+		// Cancel this transaction
+		self.inner.rollback();
+
+		// Continue
+		Ok(())
+	}
+	/// Commit a transaction
+	pub(crate) async fn commit(&mut self) -> Result<(), Error> {
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Check to see if transaction is writable
+		if !self.write {
+			return Err(Error::TxReadonly);
+		}
+		// Mark this transaction as done
+		self.done = true;
+
+		// Commit this transaction
+		self.inner.commit().await?;
+		// Continue
+		Ok(())
+	}
+	/// Check if a key exists
+	pub(crate) async fn exi<K>(&mut self, key: K) -> Result<bool, Error>
+	where
+		K: Into<Key>,
+	{
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Check the key
+		self.inner
+			.get(&key.into().as_slice())
+			.map(|opt| opt.is_some())
+			.map_err(|e| Error::Tx(format!("Unable to get kv from SurreaKV: {}", e)))
+	}
+	/// Fetch a key from the database
+	pub(crate) async fn get<K>(&mut self, key: K) -> Result<Option<Val>, Error>
+	where
+		K: Into<Key>,
+	{
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Get the key
+		let res = self.inner.get(&key.into().as_slice())?;
+
+		// Return result
+		Ok(res)
+	}
+	/// Obtain a new change timestamp for a key
+	/// which is replaced with the current timestamp when the transaction is committed.
+	/// NOTE: This should be called when composing the change feed entries for this transaction,
+	/// which should be done immediately before the transaction commit.
+	/// That is to keep other transactions commit delay(pessimistic) or conflict(optimistic) as less as possible.
+	#[allow(unused)]
+	pub(crate) async fn get_timestamp<K>(&mut self, key: K) -> Result<Versionstamp, Error>
+	where
+		K: Into<Key>,
+	{
+		// Check to see if transaction is closed
+		if self.done {
+			return Err(Error::TxFinished);
+		}
+		// Write the timestamp to the "last-write-timestamp" key
+		// to ensure that no other transactions can commit with older timestamps.
+		let key_vec = key.into();
+		let k = key_vec.as_slice();
+		let prev = self.inner.get(k.clone())?;
+
+		let ver = match prev {
+			Some(prev) => {
+				let slice = prev.as_slice();
+				let res: Result<[u8; 10], Error> = match slice.try_into() {
+					Ok(ba) => Ok(ba),
+					Err(e) => Err(Error::Ds(e.to_string())),
+				};
+				let array = res?;
+				let prev: u64 = try_to_u64_be(array)?;
+				prev + 1
+			}
+			None => 1,
+		};
+
+		let verbytes = u64_to_versionstamp(ver);
+
+		self.inner.set(k, verbytes.as_slice())?;
```

Comment:
Im wondering if we can actually expose a clock from surrealkv. Or if we want to.

The KV store does have the concept of incremental unique identifiers right? Ordering. Could have that ordering combined with system clock. But then system clock would missbehave.

Maybe we should just keep things simple like they are now until we have a specific need for this clock. Until then, the clock stuff is counter logic without underlying storage aide, iiuc.

Reply:
Actually the versioning APIs are not enabled on the KV store because I didn't see it being used for the kv backends currently. If need arises, we can definitely look into this.

Reply:
If SKV already has a notion of "write timestamp"(I believe so) internally, I wish SKV provides an API to "put a value onto a key suffixed the skv write timestamp"  similar to FDB, and we could use it from day one(this PR)!

It's a kind of "push down" that makes it possible to build arbitrary "commit log" or "change history" like data structure on top of the kvs.

Without that, we need to throttle writes due to separate "get the write timestamp of this transaction(or emulate that by storing our own time-stamp oracle state in a specific key) from the underlying kvs" and taking gap lock or advisory lock to build a linearizable commit history. Today, any SDB backend except FDB should be killing performance due to that- Every write to a CF-enabled table/database contend on reading and writing the state of our own versionstamp oracle backed by a specific key in the underlying KVS.

Reply:
That is a really good point @mumoshu . We could maybe do that in a follow up PR though. But it would be a game changer.

Reply:
I am tracking this [here](https://github.com/surrealdb/surrealkv/issues/20) @mumoshu as I'll add it as a feature in a separate PR. Thank you for highlighting this